I want to create another baseline that compares the predicted over under line to the actual line to give me a good idea of what's possible.

In [9]:
# Imports

import nflreadpy as nfl
import pandas as pd
from sklearn.metrics import mean_squared_error, mean_absolute_error

In [10]:
# Set pandas settings
pd.set_option('display.max_columns', None)

In [11]:
schedule = nfl.load_schedules(seasons = [2024, 2025])

schedule = schedule.to_pandas()

In [12]:
schedule.head(5)

game_id  season game_type  week     gameday   weekday gametime  \
0   2024_01_BAL_KC    2024       REG     1  2024-09-05  Thursday    20:20   
1   2024_01_GB_PHI    2024       REG     1  2024-09-06    Friday    20:15   
2  2024_01_PIT_ATL    2024       REG     1  2024-09-08    Sunday    13:00   
3  2024_01_ARI_BUF    2024       REG     1  2024-09-08    Sunday    13:00   
4  2024_01_TEN_CHI    2024       REG     1  2024-09-08    Sunday    13:00   

  away_team  away_score home_team  home_score location  result  total  \
0       BAL        20.0        KC        27.0     Home     7.0   47.0   
1        GB        29.0       PHI        34.0  Neutral     5.0   63.0   
2       PIT        18.0       ATL        10.0     Home    -8.0   28.0   
3       ARI        28.0       BUF        34.0     Home     6.0   62.0   
4       TEN        17.0       CHI        24.0     Home     7.0   41.0   

   overtime  old_game_id     gsis nfl_detail_id           pfr   pff  \
0       0.0   2024090500  59508.0          None  202409050kan  None   
1       0.0   2024090600  59509.0          None  202409060phi  None   
2       0.0   2024090800  59510.0          None  202409080atl  None   
3       0.0   2024090801  59511.0          None  202409080buf  None   
4       0.0   2024090802  59512.0          None  202409080chi  None   

        espn   ftn  away_rest  home_rest away_moneyline home_moneyline  \
0  401671789  6449          7          7            124           -148   
1  401671805  6450          7          7            110           -130   
2  401671744  6451          7          7            160           -192   
3  401671617  6452          7          7            250           -310   
4  401671719  6453          7          7            164           -198   

   spread_line away_spread_odds home_spread_odds  total_line under_odds  \
0          3.0             -118             -102        46.0       -110   
1          2.0             -110             -110        49.5       -112   
2          4.0             -110             -110        43.0       -115   
3          6.5             -105             -115        46.0       -112   
4          4.0             -108             -112        43.0       -110   

  over_odds  div_game      roof    surface  temp  wind  away_qb_id  \
0      -110         0  outdoors      grass  67.0   8.0  00-0034796   
1      -108         0  outdoors       None   NaN   NaN  00-0036264   
2      -105         0    closed  fieldturf   NaN   NaN  00-0036945   
3      -108         0  outdoors     a_turf  61.0  20.0  00-0035228   
4      -110         0  outdoors      grass  67.0   8.0  00-0039152   

   home_qb_id   away_qb_name     home_qb_name       away_coach  \
0  00-0033873  Lamar Jackson  Patrick Mahomes    John Harbaugh   
1  00-0036389    Jordan Love      Jalen Hurts     Matt LaFleur   
2  00-0029604  Justin Fields     Kirk Cousins      Mike Tomlin   
3  00-0034857   Kyler Murray       Josh Allen  Jonathan Gannon   
4  00-0039918     Will Levis   Caleb Williams   Brian Callahan   

       home_coach        referee stadium_id                          stadium  
0       Andy Reid  Shawn Hochuli      KAN00  GEHA Field at Arrowhead Stadium  
1   Nick Sirianni    Ron Torbert      SAO00                Arena Corinthians  
2   Raheem Morris    Brad Rogers      ATL97            Mercedes-Benz Stadium  
3  Sean McDermott      Tra Blake      BUF00                    New Era Field  
4   Matt Eberflus    Shawn Smith      CHI98                    Soldier Field

In [14]:
schedule = schedule[schedule['total_line'].notna() & schedule['total'].notna()]

In [15]:
mse = mean_squared_error(schedule['total'], schedule['total_line'])
mae = mean_absolute_error(schedule['total'], schedule['total_line'])

print(f"MSE: {mse}\nMAE: {mae}\n")

MSE: 167.04120267260578
MAE: 10.11358574610245



Vegas's best models predict the total points with a mean absolute error of 10.11 so hopefully I can get mine for predicting individual team scores around 6.